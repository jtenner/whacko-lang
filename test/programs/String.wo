export fn _start(): void {
  let a = "Hello world";
  console.log(a);
}