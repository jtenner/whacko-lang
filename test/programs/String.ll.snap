; ModuleID = 'whacko'
source_filename = "whacko"
target triple = "wasm32-wasi"

@"Hello world~484" = constant [12 x i8] c"Hello world\00"

declare ptr @__whacko_gc_alloc(i32, i32)

declare void @__whacko_gc_store_field_ptr(i32, ptr, ptr, ptr)

define void @_start() {
entry:
  %"alloca~0" = alloca ptr, align 8
  %"result~487" = call ptr @__whacko_gc_alloc(i32 1, i32 add (i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32 }, ptr null, i32 1) to i32), i32 11))
  %"gep~488" = getelementptr { ptr, ptr, i32, i32 }, ptr %"result~487", i32 1, i32 0
  call void @llvm.memcpy.p0.p0.i32(ptr %"gep~488", ptr @"Hello world~484", i32 11, i1 false)
  store ptr %"result~487", ptr %"alloca~0", align 8
  ret void
}

; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: readwrite)
declare void @llvm.memcpy.p0.p0.i32(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i32, i1 immarg) #0

declare void @__whacko_gc_visit(ptr)

define void @__visit_whacko_object(ptr %0) {
"entry~489":
  %"typeGEP~491" = getelementptr { ptr, ptr, i32, i32 }, ptr %0, i32 2
  %1 = load i32, ptr %"typeGEP~491", align 4
  switch i32 %1, label %"unreachable~490" [
    i32 1, label %"classType:[std/String.wo]String<>"
  ]

"unreachable~490":                                ; preds = %"entry~489"
  unreachable

"classType:[std/String.wo]String<>":              ; preds = %"entry~489"
  ret void
}

attributes #0 = { nocallback nofree nounwind willreturn memory(argmem: readwrite) }
